src/box/lua/net_box.c

```diff
@@ -2166,8 +2193,18 @@ luaT_netbox_transport_perform_async_request(struct lua_State *L)
 {
 	struct netbox_transport *transport = luaT_check_netbox_transport(L, 1);
 	struct netbox_request *request = lua_newuserdata(L, sizeof(*request));
-	if (luaT_netbox_transport_make_request(L, 2, transport, request) != 0)
-		return luaT_push_nil_and_error(L);
+	if (luaT_netbox_transport_make_request(L, 2, transport, request) != 0) {
+		enum netbox_method method = lua_tointeger(L, 9);
+		if (method != NETBOX_PING)
+			return luaT_push_nil_and_error(L);
+		/**
+		 * Simulate future object with nil, error as ping must
+		 * return it anyway.
+		 */
+		netbox_init_request(L, 2, request, method, 0);
+		netbox_request_set_error(request, diag_last_error(diag_get()));
+		request->transport = NULL;
```

Comment:
That seems like a huge crutch. Perhaps it comes from your decision not to remove box exceptions from `_request`. In my proposal I said that we should stop throwing box errors from `_request` and add a function `_request_ex` which will call `_request` + throw if the latter failed. This is why you have this 'future return' simulation and a `pcall` in ping code in Lua with complicated handling.

Reply:
Yeah, I agree with you, that this is a crutch. But it's not related to exceptions in `_request`. I think we should return future object in case there's no non connection related errors occured (the ones that are thrown from `_request` and not from `transport:perform_request`. If this simulated future object isn't returned the behavior of the ping will be like this: 
```lua
tarantool> c:ping({is_async = 1})
---
- null
- Peer closed
...
```
This is definitely connection related problem. But as peer was closed before the registration of the request, `luaT_netbox_transport_make_request` will return an error and no future object will be returned to user. It will be extremely non-obvious for user to determine what will be returned from async `ping` method.  Simulating future is intended to solve this problem.

As for exceptions in `_request`, now they are used to decide whether error occured inside `transport` or not. I don't see how the solution you propose (`_request` without exceptions in ping) solve the problem. We have to manually handle returned error and determine if it was connection related. This can be done with the help of setting `prev` exception when the method is ping and error is thrown from `transport` or by returning special error under the same conditions. Manual checking of error's codes can be done too but it's not a great solution. 

However, all of this doesn't solve the problem of returning nil from async ping in case of connection related issues. If it's not a great problem and we want the user to always check if async returned nil or table, we can not mess around with internals of `transport` and get by just with handling errors.

Reply:
> This is definitely connection related problem. But as peer was closed before the registration of the request, luaT_netbox_transport_make_request will return an error and no future object will be returned to user. It will be extremely non-obvious for user to determine what will be returned from async ping method. Simulating future is intended to solve this problem.

All the other future-returning methods behave like this - they fail immediately and return nil,err instead of a future. Lets keep it consistent. Async can return nil,err immediately, and it does that regularly. Ping shouldn't be an exception.